### ISSUE TITLE
Feature Request: "extract node"

### ISSUE THREAD

**Issue by tangentstorm** [leo-editor/leo-editor#3360]
Often when I'm trying to read a piece of unfamiliar code, I will import it into Leo.

When the code is very long (say a function that is several screens of text), I like to copy and paste chunks of the text into sub-nodes and replace them with `<<one of these links>>`.

It would be really nice to be able to select a span of text, right click and have an "extract node"  button that prompted for the headline text and then did this for me. (And even better if it trimmed off leading whitespace on each line of the new node!)

[I realize if I learned enough about scripting and customizing Leo I could probably do this myself, but I'm just not there yet...]

**Comment by edreamleo** [comment]
@tangentstorm Have you tried Leo's parse-body command?

**Comment by tangentstorm** [comment]
Ah, for files it can parse, that seems to do the same thing (or something very similar to) what just importing with `@clean` does.

I think probably I'm using Leo in a way that's different from most people here.

Normally I do like having leo break files into nodes for me, but often I have to deal with huge chunks of code (individual functions that are hundreds of lines long) in languages that leo doesn't know how to parse.

It does a great job of syntax-highlighting my K files, for example, but when I run parse-body in a K file, it gives me this:

```Traceback (most recent call last):
  File "C:\Python311\Lib\site-packages\leo\core\leoKeys.py", line 2507, in callAltXFunction
    func(event)
  File "C:\Python311\Lib\site-packages\leo\core\leoImport.py", line 2619, in parse_body_command
    c.importCommands.parse_body(c.p)
  File "C:\Python311\Lib\site-packages\leo\core\leoImport.py", line 1173, in parse_body
    ext = '.' + d.get(language)
          ~~~~^~~~~~~~~~~~~~~~~
TypeError: can only concatenate str (not "NoneType") to str
```

But even if Leo _could_ parse it for me, I'm not sure I'd always want it to... Part of the issue is that I might carve out several chunks that aren't right next to each other, and I'm just trying to manually carve it up in a way that lets me see the structure and give names to the bits I'm looking at:

```pascal
case whatever of
   1 : short_thing;
   2 : easy_to_read_one_liner;
   3 : <<300 lines of code>>>;
   4 : whatever;
   5 : <<another huge chunk>>;
   6 : etc
end
```

Basically, I'm doing the kind of thing I would do with a refactoring browser if I had one for the language: extracting methods or functions that mean something and giving them names. A good refactoring tool that can do that safely and figure out what parameters need to be passed in, but without that, it can be very error-prone to just start moving code around.

Leo is the next best thing to a refactoring tool, because I can move the code around and break it into chunks visually, without changing the code on disk at all.

So to recap:
- `@others` is out because the chunks are non-contiguous.
- I want to write the names manually in context anyway, because I'm trying to refactor.

**Comment by gatesphere** [comment]
The 'extract' command (default keybind ctrl+shift+d) is so close to what
you're looking for, but misses the mark by a little bit.  It works fine
when you can extract a whole function definition, but for small chunks in
the middle of a node that aren't full function bodies, or in language that
Leo doesn't have as robust support for, I'm frequently doing the same
manual thing as @tangentstorm.

**Comment by tbpassin** [comment]
I do sometimes use _extract_ on chunks in the middle of nodes.  You do need to be aware of the variables that it uses, the more so if you are going to move the chunks around as in refactoring.  I sometimes do this to hide details that interfere with my understanding of the flow of the code.  It turns these parts of the code into a kind of pseudo-code.  Other times I do it to increase the fraction of the routine's flow that I can read on a screen.

Here's a partial example where the named sections do not correspond to function definitions.  Creating the named sections was part of some refactoring I did:

```
def __init__(self, g, c):
    super().__init__()
    self.c = c
    self.g = g
    rb = self.resbrowser = QTextBrowser()
    sb = self.subjbrowser = QTextBrowser()
    relb = self.relbrowser = QTextBrowser()

    self.search_box = SearchBox(c)
    self.view_switcher = ViewSwitcher()

    self.model_data = None
    self.subj_term_data = None
    self.HTML_CSS = ''

    self.subjects = 0
    self.resources = 0

    << compute styles >>
    << set up widgets >>
    << build display widget >>
    << set layout >>
    << connect signals >>
 ```

BTW, you can use _extract_ on any selected lines, not just `<< named sections>>`  Also, CTRL-CLICKing on a named section line navigates you to that part of the code (this is a relatively recent enhancement).

I will admit that writing the outline in the way doesn't help non-Leo users who want to work with the external files outside of Leo.  To them, the files (and functions, etc) probably seem unnecessarily long.  But then, they should be using Leo(!), shouldn't they?

**Comment by edreamleo** [comment]
I am going to close this issue. Leo's existing commands should suffice.
